tenders/common/elf.c

```diff
@@ -73,120 +74,155 @@ static ssize_t pread_in_full(int fd, void *buf, size_t count, off_t offset)
+    e_entry = ehdr->e_entry;
 
-    ph_off = hdr.e_phoff;
-    ph_entsz = hdr.e_phentsize;
-    ph_cnt = hdr.e_phnum;
-    ph_size = ph_entsz * ph_cnt;
+    size_t ph_size = ehdr->e_phnum * ehdr->e_phentsize;
     phdr = malloc(ph_size);
     if (!phdr)
         goto out_error;
-    nbytes = pread_in_full(fd_kernel, phdr, ph_size, ph_off);
+    nbytes = pread_in_full(fd_kernel, phdr, ph_size, ehdr->e_phoff);
     if (nbytes < 0)
         goto out_error;
     if (nbytes != ph_size)
         goto out_invalid;
 
     /*
-     * Load all segments with the LOAD directive from the elf file at offset
-     * p_offset, and copy that into p_addr in memory. The amount of bytes
-     * copied is p_filesz.  However, each segment should be given
-     * p_memsz aligned up to p_align bytes on memory.
+     * Load all program segments with the PT_LOAD directive.
      */
-    for (ph_i = 0; ph_i < ph_cnt; ph_i++) {
-        uint8_t *daddr;
-        uint64_t _end;
-        size_t offset = phdr[ph_i].p_offset;
-        size_t filesz = phdr[ph_i].p_filesz;
-        size_t memsz = phdr[ph_i].p_memsz;
-        uint64_t paddr = phdr[ph_i].p_paddr;
-        uint64_t align = phdr[ph_i].p_align;
-        uint64_t result;
-        int prot;
+    e_end = 0;
+    for (Elf64_Half ph_i = 0; ph_i < ehdr->e_phnum; ph_i++) {
+        Elf64_Addr p_vaddr = phdr[ph_i].p_vaddr;
+        Elf64_Xword p_filesz = phdr[ph_i].p_filesz;
+        Elf64_Xword p_memsz = phdr[ph_i].p_memsz;
+        Elf64_Xword p_align = phdr[ph_i].p_align;
+        Elf64_Addr p_vaddr_end;
 
         if (phdr[ph_i].p_type != PT_LOAD)
             continue;
 
-        if ((paddr >= mem_size) || add_overflow(paddr, filesz, result)
-                || (result >= mem_size))
+        if ((p_vaddr >= mem_size)
+                || add_overflow(p_vaddr, p_filesz, p_vaddr_end)
+                || (p_vaddr_end >= mem_size))
             goto out_invalid;
-        if (add_overflow(paddr, memsz, result) || (result >= mem_size))
+        /* 
+         * Each segment should be given p_memsz aligned up to p_align bytes of
+         * memory, compute this as p_vaddr_end.
+         */
+        if (add_overflow(p_vaddr, p_memsz, p_vaddr_end)
+                || (p_vaddr_end >= mem_size))
             goto out_invalid;
         /*
-         * Verify that align is a non-zero power of 2 and safely compute
-         * ((_end + (align - 1)) & -align).
+         * Verify that p_align is a non-zero power of 2 and safely compute
+         * ((p_vaddr_end + (p_align - 1)) & -p_align).
          */
-        if (align > 0 && (align & (align - 1)) == 0) {
-            if (add_overflow(result, (align - 1), _end))
+        if (p_align > 0 && (p_align & (p_align - 1)) == 0) {
+            if (add_overflow(p_vaddr_end, (p_align - 1), p_vaddr_end))
                 goto out_invalid;
-            _end = _end & -align;
-        }
-        else {
-            _end = result;
+            p_vaddr_end = p_vaddr_end & -p_align;
         }
```

Comment:
I'm not 100% sure I'm following the bit hacking here, but it seems to me that if `p_align` is a multiple of 2 then: 
```c
Elf64_Addr  p_vaddr_end = ((p_vaddr + p_memsz) + (p_align -1)) & (-p_align);
```
It has been verified previously that `p_vaddr + p_filesz < mem_size` and that `p_vaddr + p_memsz < mem_size`, but I'm wondering if a large `p_align` can lead us change the memory protection flags of a range that goes beyond `mem_size` ?

Reply:
Good question. Also not obvious if we shouldn't be doing the `memset()` using the size adjusted for alignment. WIll check.